bbbc021_feature_reduction_tests.py

```diff
@@ -0,0 +1,247 @@
+        if len(new_subdir) != 0 and len(new_subdir_illum) != 0:
+               valid_filenames.append(new_subdir)
+               illum_filenames.append(new_subdir_illum)
+
+    return (valid_filenames, illum_filenames)
+
+
+def run_illum(filenames, names_out):
+    """
+    Find background illumination and correct all images corresponding to elements in filenames.
+
+    Save corrected images using names_out which includes a relative path from the top level directory.
+
+    :param filenames: list of valid filenames with relative paths from top level directory
+    :param names_out: list of valid filenames for saving output with relative paths from top level directory
+    """
+    illum = find_background_illumination(filenames)
+    corrected_images = correct_multiimage_illumination(filenames, illum=illum)
+    for (image, name) in zip(corrected_images, names_out):
+        io.imsave(name, image)
+
+
+def run_montage(filenames, names_out):
+    """
+    Read images from filenames and stitch and stack their quadrants and channels before saving to new files using
+    names_out
+
+    :param filenames: list of filenames with relative paths to top level sorted by well, quadrant and channel e.g.
+                        filenames = ['B02_s1_w1_illum.tif', 'B02_s1_w2_illum.tif', 'B02_s1_w4_illum.tif',
+                                    'B02_s2_w1_illum.tif', 'B02_s2_w2_illum.tif', 'B02_s2_w4_illum.tif',
+                                    'B02_s3_w1_illum.tif', 'B02_s3_w2_illum.tif', 'B02_s3_w4_illum.tif',
+                                    'B02_s4_w1_illum.tif', 'B02_s4_w2_illum.tif', 'B02_s4_w4_illum.tif']
+                        will result in one image (B02) with quadrants [[s1, s2], [s3, s4]] where each quadrant
+                        is stacked in the order [w4, w2, w1]. This example assumes files at the top level directory
+    :param names_out: list of filenames with relative paths to top level for output
+    """
+    illumed_ims = map(io.imread, filenames)
+    montaged_ims = montage_stream(illumed_ims, montage_order=[[0, 1], [2, 3]], channel_order=[2, 1, 0])
+    for (image, name) in zip(montaged_ims, names_out):
+        io.imsave(name, image)
+
+
+def output_features(ims, filenames, out_file):
+    """
+    Build a default feature map for each image in ims and output a dataframe of
+    [filenames, features] to out_file as csv for reading in
+
+    :param ims: opened nparray images
+    :param filenames: filenames corresponding to each image in ims with relative path to top level directory
+    :param out_file: name of CSV file to save dataframe, with relative path to top level directory
+    """
+    # generate filenames column to exist as first column of feature DF
+    filenames_col = ["Filenames"]
+    filenames_col.extend(filenames)
+    filenames_col = pd.DataFrame(filenames_col)
+
+    all_image_features = pd.DataFrame()
+    # set up flag to only add header row once
```

Comment:
Like I mentioned earlier (and I know you haven't had time to do anything yet), this logic with the flag is clever but unnecessary - pandas has a keyword argument "columns" for defining the header row. 

```
df = pd.DataFrame(columns=['col1', 'col2', 'col3']) . # returns an empty dataframe with column headings 'col1', 'col2', 'col3'
df = pd.DataFrame(data, columns=['col1', 'col2', 'col3'])  # returns a dataframe with data and column headings
```
https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html